### PR TITLE
directories: prevent inode cache fragmentation by orderly verifying data directory contents

### LINKED ISSUE #14506: Wasted memory in inode cache and dentry cache
An sstable is composed of 7 files. They will be opened in close proximity. Two will remain open, five will be closed.

After some time, the kernel will run out of memory. It will start evicting non-open files. But because 2 out of 7 files are still open, memory fragmentation will prevent the slabs containing those inodes/dentries from being reclaimed.

Let's take for example 200k sstables. That's 200k\*1k\*7 = 1.4GB memory used, of which 200k\*1k\*5 = 1GB is wasted.

### PR DESCRIPTION
**Pull request by lkshminarayanan**
During startup, the contents of the data directory are verified to ensure that they have the right owner and permissions. Verifying all the contents, which includes files that will be read and closed immediately, and files that will be held open for longer durations, together, can lead to memory fragementation in the dentry/inode cache.

Mitigate this by updating the verification in a such way that these two set of files will be verified separately ensuring their separation in the dentry/inode cache.

Fixes https://github.com/scylladb/scylladb/issues/14506

### CODE REVIEW

**Comment by lkshminarayanan** [comment]
From my manual testing, it looks like this patch doesn't improve the situation. The files are loaded into the dentry cache as soon as we check for permissions in the `sstable_directory::filesystem_components_lister::prepare` function call. A better fix is to probably update the prepare method to check for permissions in two phases. I'll update this PR with the new fix.

File: sstables/sstable_directory.hh [hunk 1]
```diff
@@ -170,13 +171,13 @@ private:
 
     future<> process_descriptor(sstables::entry_descriptor desc, process_flags flags);
     void validate(sstables::shared_sstable sst, process_flags flags) const;
-    future<sstables::shared_sstable> load_sstable(sstables::entry_descriptor desc, sstables::sstable_open_config cfg = {}) const;
+    future<sstables::shared_sstable> load_sstable(sstables::entry_descriptor desc, sstables::sstable_open_config cfg = {}, bool skip_loading_data_and_index = false) const;
```
Comment (denesb) [inline sstables/sstable_directory.hh:174]: Adding a new flag like this introduces a lot of churn (all callers have to be adjusted). We usually split this out into a separate commit, which just adds the flag and updates all callers, but it doesn't use it in the method yet.

Then we add the logic actually using the flag in the method in a 2nd commit. Lastly, a 3rd commit changes the interested callers to pass the non-default value of the flag. If all callers pass the same thing, than this can be folded into the 1st patch.

This amount of separation might seem superfluous, but it makes the review really easy, because reviewers can just glance over the chore patch (adding the flag) and go straight into reviewing the actually interesting changes.

Since this patch is not that big, I don't think you need to split it, just keep this in mind for future PRs.

File: sstables/sstable_directory.cc [hunk 2]
```diff
@@ -250,6 +250,13 @@ future<> sstable_directory::system_keyspace_components_lister::prepare(sstable_d
     }
 }
 
+future<> sstable_directory::load_data_and_index(process_flags flags) {
+    auto loaded_sstables = flags.sort_sstables_according_to_owner?_unshared_local_sstables:_unsorted_sstables;
```
Comment (denesb) [inline sstables/sstable_directory.cc:254]: Space around `?` and `:`.

**Comment by lkshminarayanan** [comment]
@denesb @raphaelsc Thank you for your comments! Unfortunately I had to come up a new fix as my previous PR didn't solve the issue. The entries are loaded into the inode/dentry cache as soon as we check for permissions (implemented as `access` syscall internally), so updating the loading mechanism didn't help much. 

This new fix updates the `main` method where the permissions are first checked. It treats data directory as a special case to check the permissions of its contents in two phases. Please let me know your comments in this.

File: sstables/sstable_directory.hh (hunk 1 above)
Comment (lkshminarayanan) [reply sstables/sstable_directory.hh]: Sure. I'll keep that in mind for futures PRs.

**Review by denesb — APPROVED** [review]
LGTM, I only have some administrative comments.

It is fine to have patches which just prepare the ground for the next patch. We don't squash these on merge (referring here to the 2nd patch).

The cover letter should always describe the current state of the PR, don't keep out-of-date description or changelog entries in it. These should go to separate comments below. The reason is that when merged, the top-level comment becomes the merge-commit's commit-log, and this should reflect the state of the branch as it was when merged. Changelogs and responses to review or the evolution of the PR looses meaning after the PR is merged.

@bhalevy please also review.

File: utils/directories.cc [hunk 3]
```diff
@@ -77,21 +77,20 @@ directories::directories(bool developer_mode)
...
-                        startlog.error("Failed owner and mode verification: {}", ep);
-                        return make_exception_future<>(ep);
-                    });
-                });
-            });
-        }).then([this, &locks] {
-            std::move(locks.begin(), locks.end(), std::back_inserter(_locks));
-        });
+    std::vector<file_lock> locks;
+    co_await parallel_for_each(dir_set.get_paths(), [this, &locks] (fs::path path) -> future<> {
```
Comment (bhalevy) [inline utils/directories.cc]: nit: please use coroutine::parallel_for_each. It is slightly more efficient for coroutines.

File: utils/directories.cc [hunk 4]
```diff
@@ -77,21 +77,20 @@ directories::directories(bool developer_mode)
...
-                    return directories::verify_owner_and_mode(path).handle_exception([](auto ep) {
-                        startlog.error("Failed owner and mode verification: {}", ep);
-                        return make_exception_future<>(ep);
-                    });
-                });
-            });
-        }).then([this, &locks] {
-            std::move(locks.begin(), locks.end(), std::back_inserter(_locks));
-        });
+    std::vector<file_lock> locks;
```
Comment (bhalevy) [inline utils/directories.cc:81]: nit: it would be nice to reserve the vector ahead of time if we know the dir_set size ahead of time.

Comment (bhalevy) [reply utils/directories.cc:81]: Not critical since the dir_set is small...

File: utils/directories.cc [hunk 5]
```diff
@@ -76,14 +76,14 @@ directories::directories(bool developer_mode)
         : _developer_mode(developer_mode)
 { }
 
-future<> directories::create_and_verify(directories::set dir_set) {
+future<> directories::create_and_verify(directories::set dir_set, recursive r) {
```
Comment (bhalevy) [inline utils/directories.cc:80]: `r` is not descriptive enough for this parameter.
We use single-letter names for variables that are used in smaller scope, or some we use by convention (or used to use) like `t` for table and `s` for schema.
But let's not add more.

Comment (bhalevy) [reply utils/directories.cc:80]: About the patch title and comment.
nit: needs a space between data and directory.
If the patch can be applied as stand alone without breaking anything, it's ok to keep it separate and change the note about squashing, just to mention the next step, to provide the motivation for the patch.

File: sstables/sstable_version.cc [hunk 6]
```diff
@@ -14,12 +14,14 @@ namespace sstables {
 
 const sstring sstable_version_constants::TOC_SUFFIX = "TOC.txt";
 const sstring sstable_version_constants::TEMPORARY_TOC_SUFFIX = "TOC.txt.tmp";
+const sstring sstable_version_constants::INDEX_SUFFIX = "Index.db";
+const sstring sstable_version_constants::DATA_SUFFIX = "Data.db";
```
Comment (bhalevy) [inline sstables/sstable_version.cc:18]: Nit: this change can be refactored into a standalone patch

File: utils/directories.cc [hunk 7]
```diff
@@ -104,7 +105,10 @@ void verification_error(fs::path path, const char* fstr, Args&&... args) {
 // Verify that all files and directories are owned by current uid
 // and that files can be read and directories can be read, written, and looked up (execute)
 // No other file types may exist.
-future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse, int level) {
+// If a FilterFunc is provided, verifies only the paths that return true for the filter.
+template <typename FilterFunc>
+requires std::invocable<FilterFunc, fs::path&> && std::is_same_v<std::invoke_result_t<FilterFunc, fs::path&>, bool>
```
Comment (bhalevy) [inline utils/directories.cc:110]: Why a template?  Wouldn't a std::function work just as well? (and be simpler)

File: utils/directories.cc [hunk 8]
```diff
@@ -115,6 +119,10 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
     }
     switch (sd.type) {
     case directory_entry_type::regular: {
+        if (!filterFunc(path)) {
```
Comment (bhalevy) [inline utils/directories.cc:122]: nit: coding style: we prefer camel_case over camelCase (and try not mix dufferent coding styles in the same unit)

Comment (bhalevy) [reply utils/directories.cc:122]: Also, I think it would better to give it a more descriptive name than filter_func.  Maybe do_verify, so the condition will be clearer to the reader.

File: utils/directories.cc [hunk 9]
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
 };
 
 future<> directories::verify_owner_and_mode(fs::path path, recursive recursive) {
-    return do_verify_owner_and_mode(std::move(path), recursive, 0);
+    return do_verify_owner_and_mode(std::move(path), recursive, 0, [](fs::path &path){return true;});
```
Comment (bhalevy) [inline utils/directories.cc:151]: If the signature would use a std::function parameter it should be easy to define this as the default, for backward compatibility.
and nit: please sprinkle some spaces in the lambda definition so it be more readable (like around the parenthesis and inside the curly braces)

File: utils/directories.cc [hunk 10]
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
...
+}
+
+future<> directories::verify_owner_and_mode_of_data_dir(directories::set dir_set) {
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
```
Comment (bhalevy) [inline utils/directories.cc:155]: nit: coroutine::parallel_for_each

Comment (bhalevy) [reply utils/directories.cc:155]: nit: coding style: space around parenthesis

File: utils/directories.cc [hunk 11]
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
...
+        co_await do_verify_owner_and_mode(std::move(path), recursive::yes, 0, [](fs::path &path) {
+            const sstring &file_name = path.filename().string();
```
Comment (bhalevy) [inline utils/directories.cc:157]: nit: coding style: & sticks to type.
also, you can use auto&

And also elsewhere we use native() to get to the path in a native format string  rather than string()

File: utils/directories.cc [hunk 12]
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
...
+            return file_name.ends_with(sstables::sstable_version_constants::DATA_SUFFIX) || file_name.ends_with(sstables::sstable_version_constants::INDEX_SUFFIX);
+        });
+    });
+
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
```
Comment (bhalevy) [inline utils/directories.cc:162]: nit: coroutine::parallel_for_each, space around parens

File: utils/directories.cc [hunk 13]
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
...
+        co_await do_verify_owner_and_mode(std::move(path), recursive::yes, 0, [](fs::path &path) {
+            const sstring &file_name = path.filename().string();
+            return !(file_name.ends_with(sstables::sstable_version_constants::DATA_SUFFIX) || file_name.ends_with(sstables::sstable_version_constants::INDEX_SUFFIX));
```
Comment (bhalevy) [inline utils/directories.cc:165]: this block us similar to the above that I wonder if we can encapsulate it in a reusable function

File: utils/directories.cc (hunk 4 above)
Comment (lkshminarayanan) [reply utils/directories.cc:81]: Done.

File: utils/directories.cc (hunk 7 above)
Comment (lkshminarayanan) [reply utils/directories.cc:110]: Agreed. I have changed it to a std::function.

File: sstables/sstable_version.cc (hunk 6 above)
Comment (lkshminarayanan) [reply sstables/sstable_version.cc:18]: Done

**Comment by avikivity** [comment]
How did you test it?

File: utils/directories.cc (hunk 13 above)
Comment (lkshminarayanan) [reply utils/directories.cc:165]: I have now updated this code to run in a loop to achieve the same results.

File: utils/directories.cc (hunk 12 above)
Comment (lkshminarayanan) [reply utils/directories.cc:162]: Done.

File: utils/directories.cc (hunk 11 above)
Comment (lkshminarayanan) [reply utils/directories.cc:157]: Done.

File: utils/directories.cc (hunk 9 above)
Comment (lkshminarayanan) [reply utils/directories.cc:151]: Done. An empty function is now the default and `do_verify_owner_and_mode` now checks for empty function before executing it.

File: utils/directories.cc (hunk 5 above)
Comment (lkshminarayanan) [reply utils/directories.cc:80]: Ack.

File: utils/directories.cc (hunk 8 above)
Comment (lkshminarayanan) [reply utils/directories.cc:122]: Done.

**Comment by lkshminarayanan** [comment]
@bhalevy @denesb Thank you for your reviews! I have updated the PR with a new set of commits addressing the review comments. Please let me know if this looks okay.

**Comment by lkshminarayanan** [comment]
> How did you test it?

@avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.

**Comment by avikivity** [comment]
> > How did you test it?
> 
> @avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.

Try this:

1. echo 3 > /proc/sys/vm/drop_caches, look at slab sizes (slabtop)
2. populate a few gigabytes using LeveledCompactionStrategy with sstable_size_in_mb = 1, giving a few thousand sstables
3. look at slab sizes again
4. drop_caches again (see it return to original size)
5. start the node, let it populate
6. look at slab caches, should grow by total number of files in data dir
7. drop_caches (while the node is up). look at slab caches, should show just 2 files/sstable on top of base load (in step 1)


You can repeat 6, 7 with and without the patch. Without the patch drop_caches should not materially shrink memory in use.

**Comment by avikivity** [comment]
Please adjust the subject to reflect the goal too, we're not changing the verification order for no reason.

**Comment by lkshminarayanan** [comment]
Hi @denesb @bhalevy, I have updated the PR again. The previous version didn't work as expected as the very first line of [do_verify_owner_and_mode()](https://github.com/scylladb/scylladb/blob/0cbf8f75f0163898cd6359b843f8c9d3237177f3/utils/directories.cc#L109) does a `file_stat` call and thus the inode entry gets into the cache at that point. Filtering out paths after turned out to be pointless. 

The newer version of the PR applies this filter before any of these system calls and succeeds in loading the files in the specific order.

**Comment by lkshminarayanan** [comment]
> > > How did you test it?
> > 
> > 
> > @avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.
> 
> Try this:
> 
> 1. echo 3 > /proc/sys/vm/drop_caches, look at slab sizes (slabtop)
> 2. populate a few gigabytes using LeveledCompactionStrategy with sstable_size_in_mb = 1, giving a few thousand sstables
> 3. look at slab sizes again
> 4. drop_caches again (see it return to original size)
> 5. start the node, let it populate
> 6. look at slab caches, should grow by total number of files in data dir
> 7. drop_caches (while the node is up). look at slab caches, should show just 2 files/sstable on top of base load (in step 1)
> 
> You can repeat 6, 7 with and without the patch. Without the patch drop_caches should not materially shrink memory in use.

Thank you! I was able to test the PR and the number of slabs does get reduced with the latest version of PR.

One sample from my tests :
Without the patch, total number of slabs :
1. After drop_caches but before populating the keyspaces  : 14701
2. After populating the keyspaces : 24316
3. After drop_caches : 22305

With the patch, total number of slabs :
1. After drop_caches but before populating the keyspaces  : 14401
2. After populating the keyspaces : 24212
3. After drop_caches : 16355

I can see a similar drop in the slabs used by the dentry and inode caches as well.

**Comment by denesb** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7514/)
> #### Failed Tests (3/24108):
> 
>     * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 
>     * [topology_experimental_raft.test_tablets.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_release_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.release.6)

https://github.com/scylladb/scylladb/issues/14746 (fixed already)
> 
>     * [test_token_group_based_splitting_mutation_writer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/boost/mutation_writer_test/test_token_group_based_splitting_mutation_writer) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_token_group_based_splitting_mutation_writer)
> 

https://github.com/scylladb/scylladb/issues/17076

> 
> #### Build Details:
> 
>     * Duration: 1 hr 55 min
> 
>     * Builder: spider5.cloudius-systems.com

@lkshminarayanan please rebase on current master, there are conflicts. Also, some of the above CI failures have been dealt with already.

**Comment by lkshminarayanan** [comment]
@denesb I have now rebased the PR on top of master. A recent PR moved the verification code from main into directories class. So, I have moved my fix into there as well.

**Comment by denesb** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7533/)
> #### Failed Tests (2/24144):
> 
>     * [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6234/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
> 
>     * [auth_cluster.test_password_login_message.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6234/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.1)
> 
> 
> #### Build Details:
> 
>     * Duration: 1 hr 39 min
> 
>     * Builder: spider7.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/16821

**Comment by denesb** [comment]
@avikivity @bhalevy please re-review.

**Review by avikivity — APPROVED** [review]

**Review by bhalevy — APPROVED** [review]

**Comment by lkshminarayanan** [comment]
Rebased on top of latest master.
